cmd/pebble/cmd_exec.go

```diff
@@ -0,0 +1,250 @@
+	var width, height int
+	if stdoutIsTerminal {
+		width, height, err = ptyutil.GetSize(unix.Stdout)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Start the command.
+	opts := &client.ExecOptions{
+		Command:     command,
+		Environment: env,
+		WorkingDir:  cmd.WorkingDir,
+		Timeout:     cmd.Timeout,
+		User:        user,
+		UserID:      userID,
+		Group:       group,
+		GroupID:     groupID,
+		UseTerminal: useTerminal,
+		Width:       width,
+		Height:      height,
+		Stdin:       os.Stdin,
+		Stdout:      os.Stdout,
+	}
+	if !useTerminal {
+		opts.Stderr = os.Stderr
+	}
+	process, err := cmd.client.Exec(opts)
+	if err != nil {
+		return err
+	}
+
+	// Start a goroutine to handle signals and window resizing.
+	done := make(chan struct{})
+	defer close(done)
+	go execControlHandler(process, useTerminal, done)
+
+	// Wait for the command to finish.
+	err = process.Wait()
+	switch e := err.(type) {
+	case nil:
+		return nil
+	case *client.ExitError:
+		logger.Debugf("Process exited with return code %d", e.ExitCode())
+		return &exitStatus{e.ExitCode()}
+	default:
+		return err
+	}
+}
+
+func execControlHandler(process *client.ExecProcess, useTerminal bool, done <-chan struct{}) {
+	ch := make(chan os.Signal, 10)
+	signal.Notify(ch,
+		unix.SIGWINCH, unix.SIGHUP,
+		unix.SIGTERM, unix.SIGINT, unix.SIGQUIT, unix.SIGABRT,
+		unix.SIGTSTP, unix.SIGTTIN, unix.SIGTTOU, unix.SIGUSR1,
+		unix.SIGUSR2, unix.SIGSEGV, unix.SIGCONT)
+
+	for {
+		var sig os.Signal
+		select {
+		case sig = <-ch:
+		case <-done:
+			return
+		}
+
+		switch sig {
+		case unix.SIGWINCH:
+			if !useTerminal {
+				logger.Debugf("Received SIGWINCH but not in terminal mode, ignoring")
+				break
+			}
+			logger.Debugf("Received '%s signal', updating window geometry", sig)
+			width, height, err := ptyutil.GetSize(unix.Stdout)
+			if err != nil {
+				logger.Debugf("Error getting terminal size: %v", err)
+				break
+			}
+			logger.Debugf("Window size is now: %dx%d", width, height)
+			err = process.SendResize(width, height)
+			if err != nil {
+				logger.Debugf("Error setting terminal size: %v", err)
+				break
+			}
+		case unix.SIGHUP:
+			file, err := os.OpenFile("/dev/tty", os.O_RDONLY|unix.O_NOCTTY|unix.O_NOFOLLOW|unix.O_CLOEXEC, 0666)
+			if err == nil {
+				file.Close()
+				err = process.SendSignal("SIGHUP")
+			} else {
+				err = process.SendSignal("SIGTERM")
+				sig = unix.SIGTERM
+			}
+			logger.Debugf("Received '%s' signal, forwarding to executing program", sig)
+			if err != nil {
+				logger.Debugf("Failed to forward signal '%s': %v", sig, err)
+				return
+			}
+		case unix.SIGTERM, unix.SIGINT, unix.SIGQUIT, unix.SIGABRT,
+			unix.SIGTSTP, unix.SIGTTIN, unix.SIGTTOU, unix.SIGUSR1,
+			unix.SIGUSR2, unix.SIGSEGV, unix.SIGCONT:
+			logger.Debugf("Received '%s signal', forwarding to executing program", sig)
+			err := process.SendSignal(unix.SignalName(sig.(unix.Signal)))
+			if err != nil {
+				logger.Debugf("Failed to forward signal '%s': %v", sig, err)
```

Comment:
This obviously on the nitpick camp, doubly so given that this is a debug message, but I'll mention anyway just in case. Feel free to ignore me: on snapd, pebble, and other projects, we ended up establishing "cannot" (for errors) and "Cannot" (for logs) as a standard for such messages, instead of "Error while", "Error <verb>ing", "Failed to", "Unable to", "Could not", "Can't", etc.  It's obviously not a big deal, but it's one more small detail towards a consistent overall application, that talks to the user consistently as well.  There are a few instances of that above ("Failed to forward" => "Cannot forward", "Error getting" => "Cannot get", etc).

Reply:
Done (in all except the wsutil and ptyutil libraries, which I'm leaving alone for now).